crates/sui-sdk/src/lib.rs

```diff
@@ -555,3 +558,101 @@ impl DataReader for ReadApi {
         Ok(self.get_reference_gas_price().await?)
     }
 }
+
+fn add_port_to_url(url: &str) -> Result<String, Error> {
+    let parsed_uri = reqwest::Url::parse(url)
+        .map_err(|_| Error::InvalidURI(format!("{}. Please provide an absolute URI", url)))?;
+
+    // if the URI is name.domain:port_number, then just return an error as the underlying
+    // http connection requires a scheme
+    if parsed_uri.cannot_be_a_base() {
+        return Err(Error::InvalidURI(format!(
+            "{}. URL scheme not supported, expects 'http', 'https' or 'wss'.",
+            url
+        )));
+    }
+
+    let mut port = 80;
+    if (parsed_uri.scheme() == "https" || parsed_uri.scheme() == "wss")
+        && parsed_uri.port().is_none()
+    {
+        port = 443;
+    }
+
+    let url = if parsed_uri.port().is_none() {
+        let uri = parsed_uri.to_string();
+        // the uri is usually http(s)://hostname:port/api_endpoint&params
+        // split this into "scheme+host/authority+port and the rest
+        // because we want to be able to just add the port, and then reconstruct
+        // the whole URI without messing with the scheme/host/authority and the endpoints
+        let splits = uri.splitn(4, '/').collect::<Vec<_>>();
+        // if we have an endpoint path
+        let (host, endpoint_path) = if splits.len() == 4 {
+            (splits[2], splits[3])
+        } else {
+            (splits[2], "")
+        };
+        let host = format!("{}:{}", host, port);
+        let uri = format!("{}://{}/{}", parsed_uri.scheme(), host, endpoint_path);
+        uri
+    } else {
+        parsed_uri.to_string()
+    };
```

Comment:
Similar question to the above: instead of dealing with strings, and having to take care around other parts of the URL, could we use [set_port](https://docs.rs/reqwest/latest/reqwest/struct.Url.html#method.set_port) to do the heavy lifting for us?

In this case, I don't know if we need the explicit serialization of the port, talked about in the docstring. I think not? But again, you may already be aware of limitations that I don't know about...

Reply:
Probably that explicit serialization (that I missed reading about in the docs) is what would work here. Thanks for the suggestion!

Reply:
I'm sorry for everyone I must came back set up it better.